lib/b2c/rbac_local.ts

```diff
@@ -0,0 +1,100 @@
+// This file is manually generated!
+
+import { RBACPolicy } from "./rbac";
+import { RBAC } from "./rbac";
+import { SessionsAuthorizationCheck } from "./sessions";
+import { ClientError } from "../shared/errors";
+
+// We want to refresh if the policy is more than 5 minutes old
+const MAX_AGE_MS = 1000 * 60 * 5;
+
+// PolicyCache handles local caching of RBAC policy information
+// It is based on https://github.com/panva/jose/blob/main/src/jwks/remote.ts
+// Instead of using a background worker/setInterval strategy, it refreshes on usage
+// and tracks the timestamp internally. this has a few _very nice_ properties:
+// - It works very well in Jest environments, since there's no boilerplate required to tear down the setInterval
+// - No work is done if RBAC is not used and the policy is not required
+export class PolicyCache {
+  private rbac: RBAC;
+  private _policy?: RBACPolicy;
+  private _timestamp?: number;
+
+  constructor(rbac: RBAC) {
+    this.rbac = rbac;
+  }
+
+  private fresh(): boolean {
+    return !!this._timestamp && Date.now() < this._timestamp + MAX_AGE_MS;
+  }
+
+  private async reload() {
+    const policyResponse = await this.rbac.policy();
+    this._policy = policyResponse.policy;
+    this._timestamp = Date.now();
+  }
+
+  async getPolicy(): Promise<RBACPolicy> {
+    if (!this._policy || !this.fresh()) {
+      await this.reload();
+    }
+    return this._policy as RBACPolicy;
+  }
+}
+
+export function performAuthorizationCheck({
+  policy,
+  subjectRoles,
+  authorizationCheck,
+}: {
+  policy: RBACPolicy;
+  subjectRoles: string[];
+  authorizationCheck: SessionsAuthorizationCheck;
+}): void {
+  const hasPermission = policy.roles
```

Comment:
Can we move some of this logic into `shared`? That will give us better test coverage as well since we can reuse tests across B2B and B2C